src/server/qgsserverprojectutils.cpp

```diff
@@ -314,9 +315,123 @@ QStringList QgsServerProjectUtils::wmsOutputCrsList( const QgsProject &project )
 {
-  return project.readEntry( QStringLiteral( "WMSUrl" ), QStringLiteral( "/" ), "" );
+  QString serviceUpper = service.toUpper();
+  if ( settings != NULL )
+  {
+    QString url = settings->serviceUrl( serviceUpper );
+    if ( ! url.isEmpty() )
+    {
+      return url;
+    }
+  }
+
+  QString header = QgsServerRequest::X_QGIS_SERVICE_URL;
+  if ( serviceUpper == QStringLiteral( "WMS" ) )
+  {
+    header = QgsServerRequest::X_QGIS_WMS_SERVICE_URL;
+  }
+  else if ( serviceUpper == QStringLiteral( "WFS" ) )
+  {
+    header = QgsServerRequest::X_QGIS_WFS_SERVICE_URL;
+  }
+  else if ( serviceUpper == QStringLiteral( "WCS" ) )
+  {
+    header = QgsServerRequest::X_QGIS_WCS_SERVICE_URL;
+  }
+  else if ( serviceUpper == QStringLiteral( "WMTS" ) )
+  {
+    header = QgsServerRequest::X_QGIS_WMTS_SERVICE_URL;
+  }
+  QString url = request.header( header );
+  if ( ! url.isEmpty() )
+  {
+    return url;
+  }
+  url = request.header( QgsServerRequest::X_QGIS_SERVICE_URL );
+  if ( ! url.isEmpty() )
+  {
+    return url;
+  }
+
+  QString proto;
+  QString host;
+
+  QString forwarded = request.header( QgsServerRequest::FORWARDED );
+  if ( ! forwarded.isEmpty() )
+  {
+    forwarded = forwarded.split( QLatin1Char( ',' ) )[0];
+    QStringList elements = forwarded.split( ';' );
+    foreach ( QString element, elements )
+    {
+      QStringList splited_element = element.trimmed().split( QLatin1Char( '=' ) );
+      if ( splited_element[0] == "host" )
+      {
+        host = splited_element[1];
+      }
+      if ( splited_element[0] == "proto" )
+      {
+        proto = splited_element[1];
+      }
+    }
+  }
+
+  if ( host.isEmpty() )
+  {
+    host = request.header( QgsServerRequest::X_FORWARDED_HOST );
+    proto = request.header( QgsServerRequest::X_FORWARDED_PROTO );
+  }
+
+  if ( host.isEmpty() )
+  {
+    host = request.header( QgsServerRequest::HOST );
+  }
+
+  QUrl urlQUrl = request.baseUrl();
+  if ( ! proto.isEmpty() )
+  {
+    urlQUrl.setScheme( proto );
+  }
+
+  if ( ! host.isEmpty() )
+  {
+    QStringList hostPort = host.split( QLatin1Char( ':' ) );
+    if ( hostPort.length() == 1 )
+    {
+      urlQUrl.setHost( hostPort[0] );
+      urlQUrl.setPort( -1 );
+    }
+    if ( hostPort.length() == 2 )
+    {
+      urlQUrl.setHost( hostPort[0] );
+      urlQUrl.setPort( hostPort[1].toInt() );
+    }
+  }
+
+  // https://docs.qgis.org/3.16/en/docs/server_manual/services.html#wms-map
+  QString map = request.parameter( QStringLiteral( "MAP" ) );
+  if ( ! map.isEmpty() )
+  {
+    QUrlQuery query;
+    query.setQueryItems( {{"MAP", map}} );
+    urlQUrl.setQuery( query );
+  }
+  else
+  {
+    urlQUrl.setQuery( NULL );
+  }
+  return urlQUrl.url();
+}
+
+QString QgsServerProjectUtils::wmsServiceUrl( const QgsProject &project, const QgsServerRequest &request, const QgsServerSettings *settings )
```

Comment:
Do you think that we could use a const reference for settings?

Reply:
In my last commit I try it :-)